hash.c

```diff
@@ -1745,38 +1745,48 @@ set_proc_default(VALUE hash, VALUE proc)
 /*
  *  call-seq:
  *     Hash.new                          -> new_hash
- *     Hash.new(obj)                     -> new_hash
+ *     Hash.new(default_value)           -> new_hash
  *     Hash.new {|hash, key| block }     -> new_hash
  *
- *  Returns a new, empty hash. If this hash is subsequently accessed by
- *  a key that doesn't correspond to a hash entry, the value returned
- *  depends on the style of <code>new</code> used to create the hash. In
- *  the first form, the access returns <code>nil</code>. If
- *  <i>obj</i> is specified, this single object will be used for
- *  all <em>default values</em>. If a block is specified, it will be
- *  called with the hash object and the key, and should return the
- *  default value. It is the block's responsibility to store the value
- *  in the hash if required.
- *
- *     h = Hash.new("Go Fish")
- *     h["a"] = 100
- *     h["b"] = 200
- *     h["a"]           #=> 100
- *     h["c"]           #=> "Go Fish"
- *     # The following alters the single default object
- *     h["c"].upcase!   #=> "GO FISH"
- *     h["d"]           #=> "GO FISH"
- *     h.keys           #=> ["a", "b"]
- *
- *     # While this creates a new default object each time
- *     h = Hash.new { |hash, key| hash[key] = "Go Fish: #{key}" }
- *     h["c"]           #=> "Go Fish: c"
- *     h["c"].upcase!   #=> "GO FISH: C"
- *     h["d"]           #=> "Go Fish: d"
- *     h.keys           #=> ["c", "d"]
+ *  Returns a new empty Hash object.
+ *
+ *  The initial default value and initial default proc for the new hash
+ *  depend on which form above was used.
+ *  See {Default Values}[#class-Hash-label-Default+Values].
+ *
+ *  If neither argument nor block given,
+ *  initializes both the default value and the default proc to <tt>nil</tt>:
+ *    h = Hash.new
+ *    h # => {}
+ *    h.class # => Hash
+ *    h.default # => nil
+ *    h.default_proc # => nil
+ *    h[:nosuch] # => nil
+ *
+ *  If argument <tt>default_value</tt> given but no block given,
+ *  initializes the default value to the given <tt>default_value</tt>
+ *  and the default proc to <tt>nil</tt>:
+ *
+ *    h = Hash.new(false)
+ *    h # => {}
+ *    h.default # => false
+ *    h.default_proc # => nil
+ *    h[:nosuch] # => nil
```

Comment:
This is not correct:
```ruby
h = Hash.new(false)
# => {}
h[:nosuch]
# => false
```

Reply:
Thanks, esparta, for this careful review.  Now fixed.